File: python/lib/rest.py
""" 
This file is part of the Nervatura Framework
http://nervatura.com
Copyright © 2011-2023, Csaba Kappel
License: LGPLv3
https://raw.githubusercontent.com/nervatura/nervatura/master/LICENSE 
"""

import os
import requests
import lib.utils as utils

service_url = "http://localhost:"+str(os.getenv("NT_HTTP_PORT"))+"/api"

def decodeResult(resultStr, code):
    result = utils.checkJson(resultStr)
    if type(result) == dict:
      if "code" in result:
        if result["code"] != 200 and result["code"] != 204:
          return [None, result["message"]]
    if code != 200 and code != 204:
      return [None, resultStr]
    return [result, None]

def DatabaseCreate(apiKey, options):
    headers = {"Content-Type": "application/json; charset=utf-8", "X-API-Key": apiKey }
    params = { "alias": options["database"], "demo": options["demo"] }
    try:
      response = requests.post(service_url+"/database", headers=headers, params=params)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def UserLogin(options):
    headers = {"Content-Type": "application/json; charset=utf-8" }
    try:
      response = requests.post(service_url+"/auth/login", headers=headers, json=options)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def TokenRefresh(token):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      response = requests.get(service_url+"/auth/refresh", headers=headers)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def TokenValidate(token):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      response = requests.get(service_url+"/auth/validate", headers=headers)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def UserPassword(token, options):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      response = requests.post(service_url+"/auth/password", headers=headers, json=options)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def Delete(token, options):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      path = service_url+"/"+options["nervatype"]
      if "id" in options:
        path += "?id="+options["id"]
      elif "key" in options:
        path += "?key="+options["key"]
      response = requests.delete(path, headers=headers)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def Get(token, options):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      path = service_url+"/"+options["nervatype"]
      if "ids" in options:
        if type(options["ids"]) == list:
          path += "/"+",".join(map(str, options["ids"]))
        else:
          path += "/"+options["ids"]
      else:
        if ("metadata" in options) and (options["metadata"] == True):
          path += "?metadata=true"
        else:
          path += "?metadata=false"
          if "filter" in options:
            path += "&filter="+options["filter"]
      response = requests.get(path, headers=headers)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def View(token, data):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      response = requests.post(service_url+"/view", headers=headers, json=data)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def Function(token, options):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      response = requests.post(service_url+"/function", headers=headers, json=options)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def Update(token, options):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      response = requests.post(service_url+"/"+options["nervatype"], headers=headers, json=options["data"])
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def Report(token, options):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      params = { "reportkey": options["reportkey"], "orientation": options["orientation"],
        "size": options["size"], "output": options["output"]
      }
      if "nervatype" in options:
        params["nervatype"] = options["nervatype"]
      response = requests.get(service_url+"/report?filters[@id]="+str(options["filters"]["@id"]), headers=headers, params=params)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def ReportList(token, options):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      path = service_url+"/report/list"
      if "label" in options:
        path += "?label="+options["label"]
      response = requests.get(path, headers=headers)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def ReportDelete(token, options):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      path = service_url+"/report/delete"
      if "reportkey" in options:
        path += "?reportkey="+options["reportkey"]
      response = requests.delete(path, headers=headers)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

def ReportInstall(token, options):
    headers = {
      "Content-Type": "application/json; charset=utf-8", 
      "Authorization": "Bearer "+token }
    try:
      path = service_url+"/report/install"
      if "reportkey" in options:
        path += "?reportkey="+options["reportkey"]
      response = requests.post(path, headers=headers)
      return decodeResult(response.text, response.status_code)
    except requests.ConnectionError:
      return [None, "Connection failure"]

File: php/lib/rest.php
<?php
/*
This file is part of the Nervatura Framework
http://nervatura.com
Copyright © 2011-2023, Csaba Kappel
License: LGPLv3
https://raw.githubusercontent.com/nervatura/nervatura/master/LICENSE
*/

  class Rest {
    function request($method, $path, $token, $api_key, $options){
      $service_url = 'http://'.getenv('NT_EXAMPLE_HOST').':'.getenv('NT_HTTP_PORT').'/api'.$path;
      $header = array(
        'Content-Type: application/json',
        'Accept: application/json'
      );
      if ($token != '') {
        $header = array_merge($header, array('Authorization: Bearer '.$token));
      }
      if ($api_key != '') {
        $header = array_merge($header, array('X-API-Key: '.$api_key));
      }
      $params = array(
        'method'  => $method,
        'header' =>  $header,
        'ignore_errors' => true
      );
      if ($options != NULL) {
        $data = json_encode($options);
        if ($data == NULL) {
          return array(
            'result'=>NULL,
            'error'=>'Invalid data'
          );
        }
        $params = array_merge($params, array('content' => $data));
      }

      $result = \Web::instance()->request($service_url, $params);

      if($result['headers'][2] == 'Content-Type: application/pdf'){
        return array(
          'result' => $result['body'],
          'error' => NULL
        );
      }

      $response = json_decode($result['body'], true);
      if ($response == NULL) {
        if(str_contains($result['headers'][0], '204')){
          return array(
            'result' => NULL,
            'error' => NULL
          );
        }
        return array(
          'result' => NULL,
          'error' => $result['headers'][0]
        );
      }
      if (is_array($response) && array_key_exists('code', $response) && ($response['code'] != 200) && ($response['code'] != 204)) {
        return array(
          'result' => NULL,
          'error' => $response['message'] 
        );
      }
      if(str_contains($result['headers'][0], '200') || str_contains($result['headers'][0], '204')){
        return array(
          'result' => $response,
          'error' => NULL
        );
      }
      return array(
        'result' => NULL,
        'error' => $result['headers'][0]
      );
    }

    function error($f3) {
      return array(
        'result' => NULL,
        'error' => $f3->get('ERROR.text')
      );
    }

    public function DatabaseCreate($api_key, $options) {
      $path = '/database?alias='.$options['alias'].'&demo='.$options['demo'];
      return $this->request('POST', $path, '', $api_key, $options);
    }

    public function UserLogin($options){
      return $this->request('POST', '/auth/login', '', '', $options);
    }

    public function TokenRefresh($token){
      return $this->request('GET', '/auth/refresh', $token, '', NULL);
    }
    
    public function TokenValidate($token) {
      return $this->request('GET', '/auth/validate', $token, '', NULL);
    }

    public function UserPassword($token, $options){
      return $this->request('POST', '/auth/password', $token, '', $options);
    }

    public function Delete($token, $options){
      $path = '/'.$options['nervatype'];
      if (array_key_exists('id', $options)) {
        $path = $path.'?id='.$options['id'];
      } else if (array_key_exists('key', $options)) {
        $path = $path.'?key='.urlencode($options['key']);
      }
      return $this->request('DELETE', $path, $token, '', NULL);
    }

    public function Get($token, $options){
      $path = '/'.$options['nervatype'];
      if (array_key_exists('ids', $options) && is_array($options['ids'])){
        $path = $path.'/'.implode(',', $options['ids']);
      } else if (array_key_exists('ids', $options)){
        $path = $path.'/'.$options['ids'];
      } else {
        if (array_key_exists('metadata', $options) && ($options['metadata'] == true)) {
          $path = $path.'?metadata=true';
        } else {
          $path = $path.'?metadata=false';
        }
        if (array_key_exists('filter', $options)) {
          $path = $path.'&filter='.urlencode($options['filter']);
        }
      }
      return $this->request('GET', $path, $token, '', NULL);
    }
    
    public function View($token, $data) {
      return $this->request('POST', '/view', $token, '', $data);
    }
    
    public function Function($token, $options) {
      return $this->request('POST', '/function', $token, '', $options);
    }
    
    public function Update($token, $options) {
      return $this->request('POST', '/'.$options['nervatype'], $token, '', $options['data']);
    }

    public function Report($token, $options) {
      $path = '/report?filters[@id]='.$options['filters']['@id'];
      $path = $path.'&reportkey='.$options['reportkey'];
      $path = $path.'&orientation='.$options['orientation'];
      $path = $path.'&size='.$options['size'];
      $path = $path.'&output='.$options['output'];
      if (array_key_exists('nervatype', $options)){
        $path = $path.'&nervatype='.$options['nervatype'];
      }
      return $this->request('GET', $path, $token, '', NULL);
    }

    public function ReportList($token, $options){
      $path = '/report/list';
      if (array_key_exists('label', $options)) {
        $path = $path.'?label='.urlencode($options['label']);
      }
      return $this->request('GET', $path, $token, '', NULL);
    }

    public function ReportDelete($token, $options){
      $path = '/report/delete';
      if (array_key_exists('reportkey', $options)) {
        $path = $path.'?reportkey='.urlencode($options['reportkey']);
      }
      return $this->request('DELETE', $path, $token, '', NULL);
    }

    public function ReportInstall($token, $options){
      $path = '/report/install';
      if (array_key_exists('reportkey', $options)) {
        $path = $path.'?reportkey='.urlencode($options['reportkey']);
      }
      return $this->request('POST', $path, $token, '', NULL);
    }
  }
?>